### PR TITLE
blockdb: add windowed compression with zstd

### PR DESCRIPTION
**Pull request by cce**
## Summary

Adds optional windowed-zstd compression codec for the `blkdata` and `certdata` columns of the block DB (stored in `ledger.block.sqlite`), controlled by a new `BlockDBCompressionWindow` config value, disabled by default. The codec resets every N rounds, so reads need at most N-1 prior block DB rows to decode any block; allowed values are {1, 2, 4, 8, 16, 32}. Backwards compatibility support added for switching the BlockDBCompressionWindow over time without re-encoding.

Compression performance using 20k recent mainnet rounds (60879000-60899000, 1255 MB for certs, 276 MB for blocks). Values are compressed-size ratio in %, lower is better.
```
| column | N=2  | N=4  | N=8  | N=16 | N=32 | N=64 | N=256 | N=1024 |
|--------|------|------|------|------|------|------|-------|--------|
| block  | 44.8 | 42.8 | 41.0 | 39.4 | 38.3 | 37.4 | 36.2  | 35.4   |
| cert   | 71.4 | 66.3 | 62.0 | 58.6 | 56.0 | 54.1 | 52.5  | 52.1   |
```

Block compression saturates by around N = 16; cert compression keeps gaining out to around N =32. zstd level is currently hard-coded at 11 ("best").

## Test Plan

Most new tests are in `ledger/store/blockdb/window_codec_test.go`:
   - `TestWindowCodec_Disabled` — N=0/1 round-trips through `formatRaw` verbatim.
   - `TestWindowCodec_Roundtrip` — anchor + continuation rows decode correctly across window boundaries.
   - `TestWindowCodec_RestartMidWindow` — encoder reset mid-window starts a fresh frame; decode still works via `FindAnchorOffset`.
   - `TestWindowCodec_OutOfOrder` — an out-of-order round forces a new anchor.
   - `TestWindowCodec_LegacyDetection` — pre-codec rows (no format prefix) read back unchanged.
   - `TestWindowCodec_AnchorByPrefixNotMagic` — anchor detection uses the explicit `formatWindowedAnchor` byte, not zstd magic-byte detection.
   - `TestWindowCodec_RejectsHugeUvarint` — `DecodeWindow` refuses payload uvarints above `MaxDecodedPayloadBytes` (32 MiB cap).
   - `TestMaxWindowConstantsInSync` — guards `blockdb.MaxCompressionWindow` against `config.MaxBlockDBCompressionWindow` drift.
   - `TestRoundDownRetention` — boundary cases (0, 31, 32, 33, 63, 64, 1<<40, 1<<40+31).
   - `TestEncoderPair_CloseLeavesReusable` — `Close` leaves the pair fresh for the syncer stop+start cycle.

Also some config tests... will also add a helper program to compress an existing block DB into a new sqlite file for comparison.

## Benchmarks against full 61M-round block DB

Methodology: compressed a full archive node's blockdb (61,110,166 rounds) using N=0 (no compression), 1 (zstd without window), 2, 4, 8, 16, and 32. Ran new read benchmarks to quantify read vs disk tradeoff for satisfying calls to blockdb methods like `BlockGet(rnd)`.

<details>

<summary>Full read benchmark run detail</summary>

sequential and random read block benchmarks, using c6a instance with EBS gp3 volume with default settings (3000 IOPS, 125 MB/s throughput). N=1 and N=2 benches were run separately once those compressed dests finished.

The first iteration of each `-count=3` block is the cold-cache reading; the next two are warm. Cold-cache numbers are present here for completeness but are not used as the headline below: they proved too sensitive to EBS volume state (an "Optimizing" status after a throughput change can swing cold latency 2× between runs), to the specific RNG seed's choice of round range, and to the source DB's real-world layout (built incrementally over years by a real archival node, no `BlockForgetBefore` calls, but with a different page-allocation history than the bulk-rebuilt compressed dests). Warm-cache numbers below are the mean of the 2nd and 3rd iterations and isolate per-op decompression cost.

```
goos: linux
goarch: amd64
pkg: github.com/algorand/go-algorand/ledger/store/blockdb
cpu: AMD EPYC 7R13 Processor
=== N=0  random ===
BenchmarkReader_BlockGet-32             200           3836581 ns/op
BenchmarkReader_BlockGet-32             200            151878 ns/op
BenchmarkReader_BlockGet-32             200            137117 ns/op
=== N=1  random ===
BenchmarkReader_BlockGet-32             200           2839247 ns/op
BenchmarkReader_BlockGet-32             200            195992 ns/op
BenchmarkReader_BlockGet-32             200            181358 ns/op
=== N=2  random ===
BenchmarkReader_BlockGet-32             200           5366823 ns/op
BenchmarkReader_BlockGet-32             200            203666 ns/op
BenchmarkReader_BlockGet-32             200            204813 ns/op
=== N=4  random ===
BenchmarkReader_BlockGet-32             200           5469691 ns/op
BenchmarkReader_BlockGet-32             200            258186 ns/op
BenchmarkReader_BlockGet-32             200            269240 ns/op
=== N=8  random ===
BenchmarkReader_BlockGet-32             200           7269485 ns/op
BenchmarkReader_BlockGet-32             200            352747 ns/op
BenchmarkReader_BlockGet-32             200            342270 ns/op
=== N=16 random ===
BenchmarkReader_BlockGet-32             200          10221578 ns/op
BenchmarkReader_BlockGet-32             200            548100 ns/op
BenchmarkReader_BlockGet-32             200            552318 ns/op
=== N=32 random ===
BenchmarkReader_BlockGet-32             200          14680356 ns/op
BenchmarkReader_BlockGet-32             200            871255 ns/op
BenchmarkReader_BlockGet-32             200            856222 ns/op
=== N=0  sequential ===
BenchmarkReader_BlockGetSequential-32          2000           1822419 ns/op
BenchmarkReader_BlockGetSequential-32          2000            121414 ns/op
BenchmarkReader_BlockGetSequential-32          2000            119927 ns/op
=== N=1  sequential ===
BenchmarkReader_BlockGetSequential-32          2000            410901 ns/op
BenchmarkReader_BlockGetSequential-32          2000            179904 ns/op
BenchmarkReader_BlockGetSequential-32          2000            179123 ns/op
=== N=2  sequential ===
BenchmarkReader_BlockGetSequential-32          2000            511736 ns/op
BenchmarkReader_BlockGetSequential-32          2000            196968 ns/op
BenchmarkReader_BlockGetSequential-32          2000            193618 ns/op
=== N=4  sequential ===
BenchmarkReader_BlockGetSequential-32          2000            410038 ns/op
BenchmarkReader_BlockGetSequential-32          2000            236716 ns/op
BenchmarkReader_BlockGetSequential-32          2000            239521 ns/op
=== N=8  sequential ===
BenchmarkReader_BlockGetSequential-32          2000            435003 ns/op
BenchmarkReader_BlockGetSequential-32          2000            296178 ns/op
BenchmarkReader_BlockGetSequential-32          2000            292268 ns/op
=== N=16 sequential ===
BenchmarkReader_BlockGetSequential-32          2000            526875 ns/op
BenchmarkReader_BlockGetSequential-32          2000            416122 ns/op
BenchmarkReader_BlockGetSequential-32          2000            419862 ns/op
=== N=32 sequential ===
BenchmarkReader_BlockGetSequential-32          2000            755368 ns/op
BenchmarkReader_BlockGetSequential-32          2000            681980 ns/op
BenchmarkReader_BlockGetSequential-32          2000            677915 ns/op
```

</details>

**Compression size vs steady-state read latency** (warm cache, µs/op):

| Window size (N) | Size % | Size (GB) | Sequential | Random |
|---|---:|---:|---:|---:|
| 0 (uncompressed) | 100% | 3,116 | 120 | 137 |
| 1 | 64.12% | 2,000 | 180 | 181 |
| 2 | 58.63% | 1,827 | 195 | 205 |
| 4 | 54.74% | 1,705 | 240 | 269 |
| 8 | 51.91% | 1,617 | 292 | 342 |
| 16 | 49.85% | 1,553 | 420 | 552 |
| 32 | 48.38% | 1,507 | 678 | 856 |

- N=0 is the original unmodified block DB from the archive snapshot
- Each doubling of N adds ~50-200 µs to per-op latency.

### CODE REVIEW

File: config/localTemplate.go [hunk 1]
```diff
@@ -823,6 +831,38 @@ func (e P2PHybridConfigError) Error() string {
...
 }
 
+// MaxBlockDBCompressionWindow is the largest BlockDBCompressionWindow value
+// the block database codec accepts.
+const MaxBlockDBCompressionWindow = 32
+
+// validBlockDBCompressionWindows enumerates the BlockDBCompressionWindow
+// values the codec supports. Allowed values are the divisors of
+// MaxBlockDBCompressionWindow.
+var validBlockDBCompressionWindows = []uint64{0, 1, 2, 4, 8, 16, 32}
```
Comment (jannotti) [inline config/localTemplate.go:842]: 0 means don't compress?

File: config/localTemplate.go [hunk 2]
```diff
@@ -823,6 +831,38 @@ func (e P2PHybridConfigError) Error() string {
...
+
+// GetValidatedBlockDBCompressionWindow returns the supported compression
+// window for block DB writes. Values 0 and 1 both disable compression and are
+// normalized to 1.
```
Comment (jannotti) [inline config/localTemplate.go]: Why not do a simple compression of the block individually for 1, turn off compression for 0?

File: ledger/store/blockdb/block_encoding.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+	if err != nil {
+		if err == sql.ErrNoRows {
+			err = ledgercore.ErrNoEntry{Round: rnd}
+		}
+		return
+	}
+
+	// Fast path: legacy raw rows or formatRaw rows decode without touching
+	// any other rows.
+	if !needsWindow(blkChunk) && (blkOnly || !needsWindow(certChunk)) {
```
Comment (jannotti) [inline ledger/store/blockdb/block_encoding.go]: If the current setting is windowed, perhaps do the windowed lookup first?  It may be overkill, but if it isn't, we're done in one db call. Over time, that should become the common case.

File: ledger/store/blockdb/block_encoding.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+			return
+		}
+		if !blkOnly {
+			cert, err = decodeStandaloneRow(certChunk)
+		}
+		return
+	}
+
+	// Slow path: scan up to MaxCompressionWindow rows ending at rnd. The
+	// lookback is fixed at the maximum supported window size so reads
```
Comment (jannotti) [inline ledger/store/blockdb/block_encoding.go]: Instead of just knowing `needWindow` or not, can you encode the "distance" back to the start of the window, so you're not always grabbing MaxCompressionWindow?

Or, encode the size of the window, and arrange for windows to always start at `blk%windowsSize==0`

File: ledger/store/blockdb/block_encoding.go [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+	cols := "blkdata, certdata"
+	if blkOnly {
+		cols = "blkdata"
+	}
+	var blkChunk, certChunk []byte
+	if blkOnly {
+		err = tx.QueryRow("SELECT "+cols+" FROM blocks WHERE rnd=?", rnd).Scan(&blkChunk)
+	} else {
+		err = tx.QueryRow("SELECT "+cols+" FROM blocks WHERE rnd=?", rnd).Scan(&blkChunk, &certChunk)
+	}
```
Comment (jannotti) [inline ledger/store/blockdb/block_encoding.go]: Since you needed to make this an `if/else` anyway, there's not much point in using the `cols` variable. Use the two different static queries in place.

There's not enough value in doing a prepared stmt is there?

File: ledger/store/blockdb/block_encoding.go [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+	return
+}
+
+// readWindowChunks returns the blkdata (and optionally certdata) chunks
+// for the inclusive round range [lo, hi] in ascending order. Missing rows
+// at the low end of the range are silently elided (they may legitimately
+// be missing because of round 0 or because of forget-before retention).
+// When blkOnly is true the certdata column is not fetched and the
+// returned certChunks slice is nil.
+func readWindowChunks(tx *sql.Tx, lo, hi basics.Round, blkOnly bool) (blkChunks, certChunks [][]byte, err error) {
```
Comment (jannotti) [inline ledger/store/blockdb/block_encoding.go]: If you avoid the named returns, you can `make` the slices the right size to begin with.

Even if you don't do that, can we not do a naked return?

File: ledger/store/blockdb/block_encoding.go [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+// surrounding window because the anchor chunk holds the zstd frame header
+// and the continuations hold its delta blocks.
+func needsWindow(chunk []byte) bool {
+	if len(chunk) == 0 {
+		return false
+	}
+	switch chunk[0] {
+	case formatWindowedAnchor, formatWindowedContinuation:
+		return true
+	}
```
Comment (jannotti) [inline ledger/store/blockdb/block_encoding.go]: I guess we know that an uncompressed block will start with a msgpack marker?

I was hoping we could encode the window size, or the distance to the anchor in this byte.

I this is an anchor, do you really need any more blocks? Can't you decode directly?  Or is there a problem decoding just the "prefix" of ztd stream? I see the comment claiming you need both, but I don't understand what's meant by "delta blocks" yet.

File: ledger/store/blockdb/window_codec.go [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+// percentage points of ratio without measurable encode-cost relief.
+const defaultCompressionLevel = 11
+
+// Per-row format prefix bytes for windowed rows.
+//
+// formatRaw (0x00) historically marked an uncompressed-but-prefixed row;
+// no shipping code writes it any more (disabled-codec rows are stored
+// verbatim, byte-identical to pre-codec msgp), but DecodeRaw still strips
+// it on read so any rows from an in-development build that emitted it
+// remain decodable.
```
Comment (jannotti) [inline ledger/store/blockdb/window_codec.go]: historically? Don't we just have msgpack, anchor, continuation?

**Review by jannotti — COMMENTED** [review]
I need to read more, but wanted to give some quick feedback.  It seems like we can do better by encoding the distance to the anchor in that first byte of any block.  And, instead of reading the maximum number of blocks, just read back to the anchor, and decode with zstd up to current.  I'm assuming you can convince zstd to decode properly with the prefix of the stream.

Probably we can avoid the initial read and then re-read window by always using blk%windowSize==0 as the anchor. So we have a really good guess of what we need to read right from the start.

File: ledger/store/blockdb/window_codec.go [hunk 9]
```diff
@@ -0,0 +1,388 @@
...
+// the chunk carries either formatWindowedAnchor or
+// formatWindowedContinuation as its first byte (the former for the row
+// that opens a new zstd frame, the latter for every subsequent row in
+// the same window). The returned slice is freshly allocated and safe for
+// the caller to retain.
+func (e *Encoder) EncodeRow(r basics.Round, payload []byte) ([]byte, error) {
+	if e.c.Disabled() {
+		out := make([]byte, len(payload))
+		copy(out, payload)
+		return out, nil
```
Comment (jannotti) [inline ledger/store/blockdb/window_codec.go]: Do we need a copy? If we do, prefer slices.Clone()?

File: ledger/store/blockdb/window_codec.go [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+}
+
+// MaxDecodedPayloadBytes caps the per-row decompressed payload size that
+// DecodeWindow will allocate for. A corrupt or tampered DB row could
+// otherwise encode a uvarint length close to 2^64 and cause an
+// out-of-memory panic before any error is returned. The value here is
+// generous (well above the protocol's MaxTxnBytesPerBlock and any
+// realistic certificate size) but small enough that an allocation request
+// can never lock up the process.
+const MaxDecodedPayloadBytes = 32 * 1024 * 1024 // 32 MiB
```
Comment (jannotti) [inline ledger/store/blockdb/window_codec.go]: Derive it from `bounds.MaxTxnBytesPerBlock`? (or maybe use it directly?)

Is this also used for cert sizes? What is the bound on a cert? One vote per committee member.

File: config/localTemplate.go [hunk 11]
```diff
@@ -823,6 +831,38 @@ func (e P2PHybridConfigError) Error() string {
...
+	return 0, fmt.Errorf("BlockDBCompressionWindow=%d is not supported; allowed values are %v",
+		cfg.BlockDBCompressionWindow, validBlockDBCompressionWindows)
+}
+
+// ValidateBlockDBCompressionWindow rejects unsupported values so codec write
+// paths never see settings they cannot honor.
+func (cfg Local) ValidateBlockDBCompressionWindow() error {
+	_, err := cfg.GetValidatedBlockDBCompressionWindow()
+	return err
+}
```
Comment (jannotti) [inline config/localTemplate.go]: Seems unused.

File: ledger/ledger.go [hunk 12]
```diff
@@ -117,6 +117,10 @@ func OpenLedger[T string | DirsAndPrefix](
 	log logging.Logger, dbPathPrefix T, dbMem bool, genesisInitState ledgercore.InitState, cfg config.Local,
 ) (*Ledger, error) {
 	var err error
+	cfg.BlockDBCompressionWindow, err = cfg.GetValidatedBlockDBCompressionWindow()
+	if err != nil {
+		return nil, fmt.Errorf("OpenLedger: %w", err)
+	}
```
Comment (jannotti) [inline ledger/ledger.go]: Seems like the strategy below this, when checkign the verified transaction cache's size for validity is to warn, set to a reasonable value, and continue.

I don't know which is better, but I think they should both do the same thing with bad values.

File: ledger/store/blockdb/blockdb.go [hunk 13]
```diff
@@ -58,9 +58,12 @@ func BlockInit(tx *sql.Tx, initBlocks []bookkeeping.Block) error {
 		return err
 	}
 
+	if writer == nil {
+		writer = NewBlockWriter(0)
+	}
```
Comment (jannotti) [inline ledger/store/blockdb/blockdb.go]: Any reason not to panic? Seems weird to patch up bad callers.  I think I saw it happening again at the next level down. (I don't mean to call `panic`, I mean to just proceed with the nil, no check.)

File: ledger/store/blockdb/block_encoding.go [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/ledger/ledgercore"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+func encodeBlockCertData(blk bookkeeping.Block, cert agreement.Certificate, writer *BlockWriter) (blkChunk, certChunk []byte, err error) {
+	if writer == nil {
+		writer = NewBlockWriter(0)
+	}
```
Comment (jannotti) [inline ledger/store/blockdb/block_encoding.go]: Here's the other one.  I'd just proceed without checking,

File: ledger/store/blockdb/block_encoding.go [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+	"fmt"
+
+	"github.com/algorand/go-algorand/agreement"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/ledger/ledgercore"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+func encodeBlockCertData(blk bookkeeping.Block, cert agreement.Certificate, writer *BlockWriter) (blkChunk, certChunk []byte, err error) {
```
Comment (jannotti) [inline ledger/store/blockdb/block_encoding.go]: I think it's reasonable for this (and `BlockPut`) to take pointers to the Block and Cert.  Go is moving them to the heap in each function. I guess I understand the desire to pass "values" but they both have important data in slices inside, so they're not immutable anyway.  And yet they _also_ have a reasonably large amount of normal fields, making them kind of big (especially Block).

File: ledger/store/blockdb/block_encoding.go [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+	// is found via FindAnchorOffset, so encoders that started a window
+	// mid-N (post-restart) decode correctly too.
+	lookback := uint64(MaxCompressionWindow - 1)
+	lo := rnd
+	if uint64(rnd) >= lookback {
+		lo = rnd - basics.Round(lookback)
+	} else {
+		lo = 0
+	}
+	blkChunks, certChunks, rerr := readWindowChunks(tx, lo, rnd, blkOnly)
```
Comment (jannotti) [inline ledger/store/blockdb/block_encoding.go]: You don't want to grab `rnd` again, since you already have it in `blockChunk`, `certChunk`.

File: ledger/store/blockdb/window_codec.go [hunk 17]
```diff
@@ -0,0 +1,388 @@
...
+		return false
+	}
+	switch chunk[0] {
+	case formatRaw, formatWindowedAnchor, formatWindowedContinuation:
+		return false
+	}
+	return true
+}
+
+// DecodeRaw returns the payload of a chunk that does not require windowed
```
Comment (jannotti) [inline ledger/store/blockdb/window_codec.go]: Why exported?

File: ledger/store/blockdb/window_codec.go [hunk 18]
```diff
@@ -0,0 +1,388 @@
...
+	copy(out[1:], delta)
+	return out, nil
+}
+
+// IsLegacyRaw reports whether chunk is in the pre-codec raw-msgp format.
+// Legacy rows have no format prefix; the test simply checks for the
+// absence of any of the format bytes the codec emits, taking advantage of
+// the fact that msgp-encoded blocks/certs always start with a fixmap
+// (0x80-0x8f) header.
+func IsLegacyRaw(chunk []byte) bool {
```
Comment (jannotti) [inline ledger/store/blockdb/window_codec.go]: Why exported? In fact, it can be moved into `_test.go`.

File: ledger/store/blockdb/blockdb.go [hunk 19]
```diff
@@ -343,15 +369,24 @@ func BlockEnsureSingleBlock(tx *sql.Tx) (blk bookkeeping.Block, err error) {
...
+	err = tx.QueryRow("SELECT blkdata FROM catchpointblocks WHERE rnd=?", round).Scan(&chunk)
 	if err != nil {
 		if err == sql.ErrNoRows {
 			err = ledgercore.ErrNoEntry{Round: round}
 		}
 		return
 	}
 
+	var buf []byte
+	if needsWindow(chunk) {
```
Comment (jannotti) [inline ledger/store/blockdb/blockdb.go]: If the final block is needsWindow, how can you decode it by itself?

Worse, after this function runs and prior blocks are deleted, how can _anyone_ decode it?

File: config/localTemplate.go (hunk 2 above)
Comment (cce) [reply config/localTemplate.go]: Yes, good idea, I didn't even include this as a baseline in the numbers but it does provide some good-enough savings with no window

File: ledger/store/blockdb/window_codec.go (hunk 8 above)
Comment (cce) [reply ledger/store/blockdb/window_codec.go]: oops this is some dead code related to the previous iteration I was with using vpack-style encoding that I forgot to clean up

File: ledger/store/blockdb/block_encoding.go (hunk 4 above)
Comment (cce) [reply ledger/store/blockdb/block_encoding.go]: yes over the weekend I was thinking how the prefix byte should use a few bits to encode more information like the index. The implementation already guarantees windows always start at blk % windowSize == 0

File: config/localTemplate.go (hunk 11 above)
Comment (cce) [reply config/localTemplate.go]: oops, forgot to remove, I had meant for the GetValidatedX version only to be used

File: ledger/store/blockdb/blockdb.go (hunk 19 above)
Comment (cce) [reply ledger/store/blockdb/blockdb.go]: Thinking about it more, this PR shouldn't mess with catchpoint catchup blocks (the BlockXStaging methods) at all, we download blocks out of order for performance reasons and we can just skip compression during that, and ensure you can get the compression benefits after continuing to run your node it as the block DB retention horizon deletes the initial blocks you downloaded..

File: config/localTemplate.go (hunk 1 above)
Comment (cce) [reply config/localTemplate.go:842]: Updated to make 0 mean disabled and 1 single-message compression

File: ledger/store/blockdb/block_encoding.go (hunk 3 above)
Comment (cce) [reply ledger/store/blockdb/block_encoding.go]: one DB call now

File: ledger/store/blockdb/block_encoding.go (hunk 4 above)
Comment (cce) [reply ledger/store/blockdb/block_encoding.go]: switched to using a separate column

File: ledger/store/blockdb/block_encoding.go (hunk 15 above)
Comment (cce) [reply ledger/store/blockdb/block_encoding.go]: updated to use pointers

**Comment by cce** [comment]
sqlite EXPLAIN for the read query (on an old test DB with 20K blocks in it)
```
$ sqlite3 /tmp/blockdb_n32.sqlite "EXPLAIN QUERY PLAN SELECT b.blkdata FROM blocks b, (SELECT window_start FROM blocks WHERE rnd = 425000) t WHERE b.rnd >= IFNULL(t.window_start, 425000) AND b.rnd <=
       425000 ORDER BY b.rnd ASC;"
QUERY PLAN
|--SEARCH blocks USING INTEGER PRIMARY KEY (rowid=?)
`--SEARCH b USING INTEGER PRIMARY KEY (rowid>? AND rowid<?)

$ sqlite3 /tmp/blockdb_n32.sqlite "EXPLAIN QUERY PLAN SELECT b.blkdata, b.certdata FROM blocks b, (SELECT window_start FROM blocks WHERE rnd = 425000) t WHERE b.rnd >= IFNULL(t.window_start, 425000) AND
       b.rnd <= 425000 ORDER BY b.rnd ASC;"
QUERY PLAN
|--SEARCH blocks USING INTEGER PRIMARY KEY (rowid=?)
`--SEARCH b USING INTEGER PRIMARY KEY (rowid>? AND rowid<?)
```
`rowid=?` is a B-tree primary key lookup for the inner subquery, then a primary key range scan `rowid>? AND rowid<?` for the outer. Seems performant on my laptop ... EXPLAIN doesn't say anything about needing a table scan, sort step (`ORDER BY rnd ASC` is satisfied by PK ordering), or temp tables.

Running benchmark against archive node next...
